src/test/ui/traits/object/supertrait-lifetime-bound.rs

```diff
@@ -11,6 +9,7 @@ impl<S, T: 'static> A<S> for T {}
 fn call_obj<'a>() {
     let obj: &dyn A<&'a ()> = &();
     obj.m();
+    //~^ ERROR the type `(dyn A<&'a ()> + 'static)` does not fulfill the required lifetime
```

Comment:
This... huh. So we're getting this error because, despite the `+'static` bound (which applies to the *hidden* type), the `dyn A<&'a ()>` type can't implement `A<&'a ()>` because it doesn't implement `'static`. This makes *sense*, on the one hand, but it's awfully annoying, on the other. I'm often having mixed feelings about the rules for syntactic outlives. Anyway, that's a separate issue from the one this PR is addressing, but I'd like a "cleaner" test. I wonder if this is the only way to observe the problem. Probably.

Reply:
Cleaner test:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c8a099474e10fd5ee9ed8e7888b7f046

Reply:
@nikomatsakis, that test you provided passes with this code. 

Calling `test1` with `T=dyn Bar<&'a u32>` doesn't require that the `T` parameter implements `Bar`. I think we need to add a `T: Bar` bound to `test1` like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=711903c31fa6b4e2644ed974212032cd
```rust
trait Foo: 'static { }

trait Bar<T>: Foo { }

fn test1<T: ?Sized + Bar<S>, S>() { }

fn test2<'a>() {
    // Here: the type `dyn Bar<&'a u32>` references `'a`,
    // and so it does not outlive `'static`.
    test1::<dyn Bar<&'a u32>, _>();
    //~^ ERROR the type `(dyn Bar<&'a u32> + 'static)` does not fulfill the required lifetime
}

fn main() { }
```
Which correctly fails to compile with this PR.

...Unless I misunderstood your "cleaner test" comment, and you were linking something that should pass with `// check-pass`. 
In that case, a simpler "passing" test (until we add these checks to WF) would be: `let x: Option<&dyn Bar<&'a u32>> = None;` for the method body of `test2`. 

Thoughts?

Reply:
^^ cc @nikomatsakis